extra_proposal/proposal.py

```diff
@@ -0,0 +1,463 @@
+            #  https://stackoverflow.com/questions/3429510/pythonic-way-to-convert-a-list-of-integers-into-a-string-of-comma-separated-rang
+
+            # relevant to DAMNIT
+            sequence = np.unique(sequence)
+
+            grouped_sequence = (
+                list(x) for _, x in groupby(sequence, lambda x, c=count(): next(c) - x)
+            )
+            return (
+                ",".join(
+                    "-".join(map(str, (gi[0], gi[-1])[: len(gi)]))
+                    for gi in grouped_sequence
+                ),
+                sequence.size,
+            )
+
+        print(
+            "Proposal {} ── scientific instrument {}".format(
+                self._proposal_number, self.instrument
+            )
+        )
+        print(f"Data stored at {self._proposal_directory}")
+
+        # title
+        if self.title is not None:
+            print("'{}'".format(self.title))
+
+        # runs
+        runs = self.runs()
+        print("\nRuns collected: {} (total {})".format(*run_ranges(runs)))
+        if self.damnit is not None:
+            grouped_sequence, size = run_ranges(self.damnit.runs())
+            print(
+                " └── {:.1f}% processed by DAMNIT".format(100 * size / len(runs)),
+                end="",
+            )
+            if size != len(runs):
+                print(": {} (total {})".format(grouped_sequence, size))
+            else:
+                print(".")
+
+    def search_source(
+        self, regex: str, run: Optional[int | list[int]] = None
+    ) -> dict[list[str]]:
+        """Perform a case insensitive search of the regex in all data sources and aliases.
+
+        Args:
+            regex (str): The regular expression.
+            run (Optional[int  |  list[int]], optional): Specific run or list of runs. Defaults to None.
+
+        Returns:
+            list[list[str]]: A dictionary of runs, and for each a list of strings matching the regular expression.
+
+        Example:
+            proposal = Extra(1234)
+            run = proposal[42]
+
+            run.search_sources("energy")
+        """
+
+        # TODO: look into aliases
+
+        run_list = []
+        if run is None:
+            run_list = self.runs()
+        elif type(run) is int:
+            run_list = [run]
+        elif type(run) is list:
+            run_list = run
+            if not all(type(i) is int for i in run):
+
+                logger.error("Each entry in the list must be an integer (run number)")
+                raise TypeError
+        else:
+            logger.error("{} is not supported".format(type(run)))
+            raise TypeError
+
+        run_match = {}
+        for ri in run_list:
+            run_match[ri] = []
+            for si in self[ri][0].all_sources:
+                if re.search(regex, si, re.IGNORECASE):
+                    logger.info("{}".format(si))
+
+                    run_match[ri].append(si)
+
+        return run_match
+
+    def a(self, b):
```

Comment:
Shall we rename this? :sweat_smile: I'm guessing it prints out the detector sources?